### PR TITLE
Increase the specificity of the logging rights.

### PR DESCRIPTION
**Pull request by erikerikson**
## What did you implement:
Implement a least-privilege logging rights statement in the default IAM policy.

## How did you implement it:

After a number of iterations on this PR, narrowing the logging rights of the deployed lambdas has boiled down to:

1. Remove the log group creation IAM right from the default Lambda execution policy - create the log group using the CFT instead of expecting the Lambda(s) to implicitly create it/them.
2. Rather than have a wild-card-style declaration of the resources that the lambdas are allowed to write into, specify the exact LogGroups that any of the lambdas will be writing into as an array of resources.

## How can we verify it:

Added tests to verify and confirm correct addition of the LogGroup(s)
Add tests to verify and confirm correct addition/specification of LogGroup(s) as the Resource(s) to which the Lambdas have CreateLogStream and PutLogEvents rights upon.

Beyond this programmatic verification, full verification will require deployment of a service and the verification that:

1. logging still works from the lambda's console.log statements into CloudWatch
2. that functions cannot `PutLogEvents` (say, via explicit coding against the CloudWatch API) into any CloudWatch logs not listed in the policy

Apologies but please note that I cannot deploy policies and as such cannot easily verify these things myself.

## Todos:

- [x] Write tests
- [x] Write documentation // this change doesn't impact docs, it simply increases rights specificity
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Leave a comment that this is ready for review once you've finished the implementation

### CODE REVIEW

**Comment by flomotlik** [comment]
Yup this change definitely makes sense. We want to reduce permissions that are deployed for specific roles in general and allow users to point to pre-defined roles as well if they are available. This is definitely a good start.

**Comment by harakiro** [comment]
The changes in the pull request don't seem to mirror what is being suggested here. The PR doesn't include the reference to the lambda function name.

**Comment by erikerikson** [comment]
Agreed @harakiro.  I'd prefer that it did but given the standard rights management approach of Serverless, wherein it uses the same role for all lambdas, it wouldn't work to make it include the function name.  In order to support that, each lambda would need to have its own role.  We've been discussing that option elsewhere.  Thanks for reviewing!

**Comment by erikerikson** [comment]
I have just noted that a similar thing is hard coded into [index.js#L59](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/functions/index.js#L59) except that it doesn't inject the account ID or `log-group:/aws/lambda/` into the resource.  I'd prefer this approach for its greater specificity but also because it doesn't spread the default across a config file and code in combination.  This empowers you to just look and know without the tribal knowledge of what all composes to give a result.  Also, there's no gotcha in changing one but forgetting the other.  I will add the removal of that modification.

**Comment by erikerikson** [comment]
From the latest commit:
```
Yeesh.  Sorry for the noob moves.  Resolves lint.
Note that the reduction in coverage is a result of the reduction of lines of code via removal of those lines without removal of uncovered lines.
```

**Comment by erikerikson** [comment]
Incidentally, if the strategy being advocated here is undesired, I have no problem closing the PR.  Whatever is best for SLS.  :)

**Comment by andymac4182** [comment]
It would be great if this could be implemented. As part of fixing https://github.com/serverless/serverless/pull/1934 I had also included moving https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/compile/functions/index.js#L59 to use the Cloudformation variables which is a less error prone way of implementing this. AccountId helps as well. As this is the default policy we could change this to be an array of the log groups based on the function names. This would allow for the tight locking down of the IAM policy.

File: lib/plugins/aws/deploy/compile/functions/tests/index.js [hunk 1]
```diff
@@ -114,12 +114,17 @@ describe('AwsCompileFunctions', () => {
...
-        }-${
-          awsCompileFunctions.serverless.service.service
-        }-lambda`
+      ).to.deep.equal(
+        {
+          'Fn:Join': [
+            '-',
+            [
+              '${opt:stage}',
+              '${self:service}',
```
Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/functions/tests/index.js]: These variables will be rendered by @eahefnawy's code prior to deployment, right?

**Comment by erikerikson** [comment]
> As this is the default policy we could change this to be an array of the log groups based on the function names. This would allow for the tight locking down of the IAM policy.

Slick idea.  I like it.  Thanks @andymac4182.

What do you think @flomotlik?

**Comment by andymac4182** [comment]
Not sure if this is relevant here or in its own issue but it would be great if we were defining the log groups as part of the Cloudformation template. That would allow the use of refs here and also ensure they got cleaned up as part of the remove operation. They are also required to be created by CF if we want to support things like Cloudwatch Log events from lambdas in the same service.

File: lib/plugins/aws/deploy/compile/functions/tests/index.js (hunk 1 above)
Comment (flomotlik) [inline lib/plugins/aws/deploy/compile/functions/tests/index.js]: No, the variables are only populated after loading of the service once. So adding more into the CF template doesn't work. What we can do though is putting those into the CF template as variables so they can be referenced.

**Comment by flomotlik** [comment]
Sorry for the delay on this guys. The main problem with creating the log groups automatically would be if AWS chooses to change the default name of the log groups. But I do think this is probably unlikely and we can add the log groups to the CF template directly and then remove them as well (thanks @andymac4182 , didn't even think about this before)

So I think the best implementation would be to automatically add the log groups and then referencing the functions and log groups accordingly so only they can log. Makes sense to you guys?

File: lib/plugins/aws/deploy/compile/functions/tests/index.js (hunk 1 above)
Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/functions/tests/index.js]: Thanks @flomotlik!  I'll do that so this isn't wrong

**Comment by erikerikson** [comment]
Sounds smart @flomotlik.  Happy to do.

**Comment by erikerikson** [comment]
Thanks especially to @andymac4182 for the good idea

File: lib/plugins/aws/deploy/compile/functions/tests/index.js (hunk 1 above)
Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/functions/tests/index.js]: I wonder, however, why aren't variables populated with every SLS file load?

**Comment by erikerikson** [comment]
Sorry.  Failed to update tests... incoming...

**Comment by erikerikson** [comment]
Still need to add tests verifying the log group resource additions

**Comment by erikerikson** [comment]
Okay, so that should be ready once again.  Will update PR summary.

!Full Disclosure: I cannot test this as I cannot use the separated-policy approach!

**Comment by erikerikson** [comment]
Should be actually ready for review and merging at this point

File: lib/plugins/aws/deploy/compile/functions/index.js [hunk 2]
```diff
@@ -183,6 +176,42 @@ class AwsCompileFunctions {
...
+
+      if (typeof this.serverless.service.provider.iamRoleARN !== 'string') { // only if default IAM
+        const logRetentionInDays = 14; // TODO make configurable?
+        const logGroupTemplate = `
+          {
+            "${normalizedFunctionName}LogGroup": {
+              "Type" : "AWS::Logs::LogGroup",
+              "Properties" : {
+                "LogGroupName" : "/aws/lambda/${FunctionName}",
+                "RetentionInDays" : "${logRetentionInDays}"
```
Comment (flomotlik) [inline lib/plugins/aws/deploy/compile/functions/index.js]: I don't think we should set a log retention policy unless its configurable. Otherwise that might be a surprise for people.

**Comment by flomotlik** [comment]
Nice Work Erik on this. Imho we should either remove the LogRetention Config, or make it configurable (e.g. provider->logRetentionInDays)

File: lib/plugins/aws/deploy/compile/functions/index.js (hunk 2 above)
Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/functions/index.js]: Good call.  It wasn't clear if it was an optional attribute given that no default value is specified.  A problem with not being able to do this in the native fashion.  Maybe mostly a problem with my laziness.

File: lib/plugins/aws/deploy/compile/functions/index.js [hunk 3]
```diff
@@ -183,6 +176,40 @@ class AwsCompileFunctions {
...
+
+        this.serverless.service.provider.compiledCloudFormationTemplate
+            .Resources
+            .IamPolicyLambdaExecution
+            .Properties
+            .PolicyDocument
+            .Statement[0]
+            .Resource
+            .push({
+              'Fn:Join': [
```
Comment (flomotlik) [inline lib/plugins/aws/deploy/compile/functions/index.js]: There is a typo here: "Fn:Join" -> "Fn::Join" (same in tests).

**Comment by andymac4182** [comment]
@flomotlik That error happens if you are deploying an update to an existing one that has already created logs.

File: lib/plugins/aws/deploy/compile/functions/index.js (hunk 3 above)
Comment (erikerikson) [inline lib/plugins/aws/deploy/compile/functions/index.js]: Sorry, this is exactly the problem with developing when not directly verifying.

This is changed in code and tests.

**Comment by erikerikson** [comment]
Sorry for this @eahefnawy.  This PR is waiting for my being able to set up to actually use it.  I think I've made the space but feel free to ignore it until I comment again to the contrary.  Thanks!

**Comment by eahefnawy** [comment]
no worries @erikerikson 😊  ... it's a great step toward nonetheless 💯

**Comment by erikerikson** [comment]
Okay, so I've finally got set up for use and figured out what the issue is.  The rights to create and write need to be broken out and more specifically specify the targets:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "logs:CreateLogStream"
            ],
            "Resource": [
                "arn:aws:logs:us-west-2:${accountId}:log-group:/aws/lambda/${lambdaName}"
            ],
            "Effect": "Allow"
        },
        {
            "Action": [
                "logs:PutLogEvents"
            ],
            "Resource": [
                "arn:aws:logs:us-west-2:${accountId}:log-group:/aws/lambda/${lambdaName}:*"
            ],
            "Effect": "Allow"
        }
    ]
}
```
Fix incoming.

**Comment by erikerikson** [comment]
This was manually tested with the sequence:
```
slsDev deploy -v
slsDev invoke -f hello
slsDev logs -f hello --startTime 30m
```

**Comment by erikerikson** [comment]
So that I don't leave it implicit...

This is ready for merge

**Comment by flomotlik** [comment]
Awesome @erikerikson 

@eahefnawy could you review and make sure that after serverless remove everything is properly removed. GTM from my side, when validated by you @eahefnawy we can merge.

**Comment by eahefnawy** [comment]
@erikerikson thanks for fixing it ... works beautifully 😊 ... the only issue now is that this is a breaking change 😞 . I just tried deploying one of my old services with this change included, but I wasn't able to deploy because this change tries to create a log group of the same name as the default log group created implicitly by lambda before. So I get the following error:

```
     An error occurred while provisioning your stack: HelloLogGroup
     - /aws/lambda/logging-rights-serverless-dev-hello already
     exists.

```

Even removing the old service and redeploying won't fix it because the old log group won't be removed as part of the stack. So the only fix for the user is to manually delete the log group from the aws console.

@erikerikson could you think of a way to make this change backward compatible? I thought maybe change the log group name, but I guess the log group must have this format `/aws/lambda/<lambdaName>` for lambda to pick it up. I guess deleting the old log group via the API is also viable, but that means the user will lose all of his logs.

Any other thoughts? cc/ @flomotlik

**Comment by eahefnawy** [comment]
btw @flomotlik just confirmed that the log groups gets removed as part of the stack removal. This should make cleaning up much easier as we won't hit the 500 log groups limit (again 😅 )

**Comment by erikerikson** [comment]
Thanks @eahefnawy.  This is an issue for sure and It's a bit of a sticky wicket.

Is it best that the user get an error (we could detect this case and add a useful link to an issue on the matter as well as instructions on how to resolve it) and pointers on how to fix it?
Alternatively, is it best that the framework solve for the problem and automatically wipe out user data without notice?

I suppose that it wouldn't be too hard to detect the case and ask the user whether they would like their log group and logs deleted for them.  This (or just detecting the error and adding an informative message) would produce a bit of legacy that would probably hang around for a while.

Given that the framework is still in RC status, I can see this going any of these ways.  Don't feel that it is my call.

**Comment by erikerikson** [comment]
@flomotlik - I was hoping you or the team would have an opinion on how that be done.   This isn't a how to implement but a what to implement.

**Comment by flomotlik** [comment]
@erikerikson yup sorry added the approval only so that we don't merge accidentally (wasn't meant as a specific message).

Imho we can't release this as a breaking change. Thought about releasing with 1.0 as a breaking change but even that was imho not an option because it would disrupt too many users.

I've looked into CF, but I can't seem to find an option that would allow a resource to be only created if it isn't already there. 

What is imho the best way forward now is to put this behind a variable at first, only if this variable is set to true add logs to CF. We'll add a warning (separate PR that we will do including docs) to it as well and then in the next release we document this and tell people to change it.

In 2 releases we make this behaviour mandatory, remove the option and force users into setting it up through CF.

```
provider:
  cf_logs: true
```

We'll add cf_logs: true to all templates (separate PR) and we'll remove the IAM rights to create the log group (as you've already done), so if users remove the logs the Lambda function can't recreate it.

This means users can still deploy things and the logs will totally work, but they get a message and we'll remove this in the future and they can easily move their services over.

Thoughts?

**Comment by erikerikson** [comment]
I am /too slowly/ working on detecting this case and providing users with options to address it.

**Comment by erikerikson** [comment]
oh, hadn't refreshed before posting last...

**Comment by erikerikson** [comment]
TL;DR: Good call but more complexity that way

NP on the approval, no adverse message read.

The choice not to introduce a breaking change seemed right.

What you suggest will create a bit more complexity and legacy for maintainers.  If the addition of `cf_logs: true` is implemented in a separate PR, then it will be important that the two PRs be pushed to npm and master together (or one merged onto the other before merge) so that no one ends up in a half-state (having `cf_logs` but LogGroups created by Lambda).

An alternative is to detect the error message from CF and offer to do the deletion, providing a description of the consequences with a link to an issue for more information.  If the user agrees to deletion, do the delete and run the deploy again continuing to use new logic (which should work at that point).  If the user does not agree to deletion, run the deploy again and use old logic.  In non-interactive mode the message could be written to logs as a warning and treated as a user declination.  This would create longer deploys for pre-change projects.  Still has the legacy management issue but avoids the `cf_logs` artifact and how to manage its removal.  It would be possible to force the migration (and reduce the code complexity) by not allowing users to use the old logic in the future.

No a painless, satisfying answer to this for me.

**Comment by flomotlik** [comment]
Sorry I might not have been clear, the `cf_logs: true` should be done in this PR I think. So that with this PR we're able to set the parameter and change the way logs are deployed. We'll do documentation and a warning message in a separate PR so we don't blow up this PR unnecessarily.

Would be great if you could add that parameter in this PR, we're happy to do it as well if you want.

> An alternative is to detect the error message from CF and offer to do the deletion

One general principle of mine for building serverless is that I don't want to let people run into unforeseen error states, because it always creates confusion, no matter how good the documentation around it is (we already thit that unfortunately with RCs, but we won't do it after the stable release). Thats why I think letting the error happen and ask them is a no-go. I also want users to have a little more time to prepare and see things like this coming so there isn't that one day when all deployments fail.

Another issue with anything that is interactive is CI. If deployments fail in CI because of interactive questions that creates a large amount of confusion because CI systems make interactivity kind of hard to spot sometimes (We've had this issue a Million times at Codeship).

**Comment by erikerikson** [comment]
Sorry I misunderstood @flomotlik.  Thanks for clarifying.  Happy to add `cf_logs` but if you want it don't on a faster timeline than I can push out, I won't be insulted in the least.

To clarify, I was suggesting both that non-interactive-mode would be treated specially and that the error case would initially be swallowed and but yeah, it's a tricky.  Hopefully as I work this I'll get an epiphany.

**Comment by erikerikson** [comment]
"be swallowed and **automatically remediated** but yeah, it's a tricky **circumstance to paper over**."

**Comment by flomotlik** [comment]
@erikerikson we're going to pull this in a branch and finish this up quickly today so we can merge. Thanks for all the help on this one, looking forward to chat tomorrow!

**Comment by erikerikson** [comment]
Thanks @flomotlik!  Sorry I've been so distracted on these.  Very much looking forward to meeting!
